src/main/java/nextstep/subway/line/section/Sections.java

```diff
@@ -29,41 +30,93 @@ public List<Section> get() {
         return sections;
     }
 
-    public void addSection(Section section) {
+    public List<Station> allStations() {
+        return sections.stream().flatMap(section -> section.stations().stream()).distinct().collect(Collectors.toList());
+    }
+
+    private Section firstSection() {
+        return sections.get(0);
+    }
+
+    private Section lastSection() {
+        return sections.get(sections.size() - 1);
+    }
+
+    public void addSection(Section newSection) {
         if (sections.isEmpty()) {
-            sections.add(section);
+            sections.add(newSection);
             return;
         }
 
-        if (!isLastDownStation(section.getUpStation())) {
-            throw new LineException(ErrorCode.CANNOT_ADD_SECTION, "노선 등록 시 상행역은 현재 하행 종점역이어야 합니다.");
+        if (isDuplicatedSection(newSection)) {
+            throw new LineException(ErrorCode.CANNOT_ADD_SECTION, "이미 등록된 구간입니다.");
         }
 
-        if (isDuplicatedStation(section)) {
-            throw new LineException(ErrorCode.CANNOT_ADD_SECTION, "이미 구간에 등록된 역입니다.");
+        if (isMiddleSection(newSection)) {
+            addMiddleSection(newSection);
+            return;
+        }
+
+        if (isFirstSection(newSection)) {
+            addFirstSection(newSection);
+            return;
         }
 
-        sections.add(section);
+        if (isLastSection(newSection)) {
+            addLastSection(newSection);
+            return;
+        }
+
+        throw new LineException(ErrorCode.CANNOT_ADD_SECTION, "등록할 수 없는 구간입니다.");
     }
 
-    private boolean isDuplicatedStation(Section section) {
-        return allStations().contains(section.getDownStation());
+    private boolean isDuplicatedSection(Section newSection) {
+        return sections.stream().anyMatch(section -> section.matchStations(newSection));
     }
 
-    public List<Station> allStations() {
-        return sections.stream().flatMap(section -> section.stations().stream()).distinct().collect(Collectors.toList());
+    private void addMiddleSection(Section newSection) {
+        Section originalSection = sections.stream()
+                .filter(section -> section.getUpStation().equals(newSection.getUpStation()))
+                .findFirst()
+                .orElseThrow();
+
+        int index = sections.indexOf(originalSection);
+        sections.add(index - 1 == -1 ? 0 : index - 1, newSection);
```

Comment:
아래와 같이 더 간소화 할 수 있겠네요!
```suggestion
        sections.add(index == 0 ? 0 : index - 1, newSection);
```